database/pdb.ksy

```diff
@@ -0,0 +1,961 @@
+      sometimes by two or three."
+  - contents: [0, 0, 0, 0]
+  - id: tables
+    type: table
+    repeat: expr
+    repeat-expr: num_tables
+    doc: |
+      Describes and links to the tables present in the database.
+
+types:
+  table:
+    doc: |
+      Each table is a linked list of pages containing rows of a single
+      type. This header describes the nature of the table and links to
+      its pages by index.
+    seq:
+      - id: type
+        type: u4
+        enum: page_type
+        doc: |
+          Identifies the kind of rows that are found in this table.
+      - id: empty_candidate
+        type: u4
+      - id: first_page
+        type: page_ref
+        doc: |
+          Links to the chain of pages making up that table. The first
+          page seems to always contain similar garbage patterns and
+          zero rows, but the next page it links to contains the start
+          of the meaningful data rows.
+      - id: last_page
+        type: page_ref
+        doc: |
+          Holds the index of the last page that makes up this table.
+          When following the linked list of pages of the table, you
+          either need to stop when you reach this page, or when you
+          notice that the `next_page` link you followed took you to a
+          page of a different `type`.
+    -webide-representation: '{type}'
+
+  page_ref:
+    doc: |
+      An index which points to a table page (its offset can be found
+      by multiplying the index by the `page_len` value in the file
+      header). This type allows the linked page to be lazy loaded.
+    seq:
+      - id: index
+        type: u4
+        doc: |
+          Identifies the desired page number.
+    instances:
+      body:
+        doc: |
+          When referenced, loads the specified page and parses its
+          contents appropriately for the type of data it contains.
+        io: _root._io
+        pos: _root.len_page * index
+        size: _root.len_page
+        type: page
+
+  page:
+    doc: |
+      A table page, consisting of a short header describing the
+      content of the page and linking to the next page, followed by a
+      heap in which row data is found. At the end of the page there is
+      an index which locates all rows present in the heap via their
+      offsets past the end of the page header.
+    seq:
+      - contents: [0, 0, 0, 0]
+      - id: page_index
+        doc: Matches the index we used to look up the page, sanity check?
+        type: u4
+      - id: type
+        type: u4
+        enum: page_type
+        doc: |
+          Identifies the type of information stored in the rows of this page.
+      - id: next_page
+        doc: |
+          Index of the next page containing this type of rows. Points past
+          the end of the file if there are no more.
+        type: page_ref
+      - type: u4
+        doc: |
+          @flesniak said: "sequence number (0->1: 8->13, 1->2: 22, 2->3: 27)"
```

Comment:
Could you get this field an `id` and a better `doc`? I don't understand currently (without reading the `doc-ref`) how 6his field is related to the mentioned mapping.

Reply:
No, I can’t. There is no official documentation.

Reply:
>There is no official documentation.

It doesn't matter, you can call it like `unkn` and put into the docs your guesses and observations.

Reply:
I don’t have guesses or observations about the ones I have unlabeled, and the style guide said to not give IDs to things that are unknown, which is why I took this approach. We might be able to get more ideas from @flesniak whose [Python implementation](https://github.com/flesniak/python-prodj-link) was my main source of information, but these mappings are a small part of a large project, and I have other more pressing user needs to focus on than fleshing out details beyond the ones I need for supporting stage shows, so to be honest I am not likely to have time to pursue such details. I submitted what I know as a courtesy to the KSC project since it might be useful to others with similar needs, but if this is not thorough enough, you can delete the merge request. People looking for such capabilities are frankly more likely to find my projects before they find KSC.